packages/SwingSet/src/vats/comms/dispatch.js

```diff
@@ -73,49 +73,54 @@ export function buildCommsDispatch(syscall) {
-    // get a bogus dispatch.notify. Currently we throw an error, which
-    // is currently ignored but might prompt a vat shutdown in the future.
+      // todo: if we previously held resolution authority for this promise, then
+      // transferred it to some local vat, we'll have subscribed to the kernel
+      // to hear about it. If we then get the authority back again, we no longer
+      // want to hear about its resolution (since we're the ones doing the
+      // resolving), but the kernel still thinks of us as subscribing, so we'll
+      // get a bogus dispatch.notify. Currently we throw an error, which is
+      // currently ignored but might prompt a vat shutdown in the future.
 
-    // TODO: The following goofiness, namely taking apart the capdata object and
-    // looking at it to see if it's a single presence reference and then
-    // treating it in a special way if it is, is a consequence of an impedance
-    // mismatch that has grown up between the comms protocol and the
-    // evolutionary path that the kernel/vat interface has taken.  In the future
-    // we should clean this up and unify presences and data in the comms
-    // protocol the way the kernel/vat interface has.
-    const unser = JSON.parse(data.body);
-    let resolution;
-    if (rejected) {
-      resolution = harden({ type: 'reject', data });
-    } else if (
-      Object(unser) === unser &&
-      QCLASS in unser &&
-      unser[QCLASS] === 'slot'
-    ) {
-      const slot = data.slots[unser.index];
-      insistVatType('object', slot);
-      resolution = harden({ type: 'object', slot });
-    } else {
-      resolution = harden({ type: 'data', data });
+      // TODO: The following goofiness, namely taking apart the capdata object
+      // and looking at it to see if it's a single presence reference and then
+      // treating it in a special way if it is, is a consequence of an impedance
+      // mismatch that has grown up between the comms protocol and the
+      // evolutionary path that the kernel/vat interface has taken.  In the
+      // future we should clean this up and unify presences and data in the
+      // comms protocol the way the kernel/vat interface has.
+      const unser = JSON.parse(vp.data.body);
+      let resolution;
+      if (vp.rejected) {
+        resolution = harden({ type: 'reject', data: vp.data });
+      } else if (
+        Object(unser) === unser &&
+        QCLASS in unser &&
+        unser[QCLASS] === 'slot'
+      ) {
+        const slot = vp.data.slots[unser.index];
+        insistVatType('object', slot);
+        resolution = harden({ type: 'object', slot });
+      } else {
+        resolution = harden({ type: 'data', data: vp.data });
+      }
+      resolveFromKernel(vpid, resolution);
     }
-    resolveFromKernel(promiseID, resolution);
+    // XXX question: do we need to call retirePromiseIDIfEasy (or some special
+    // comms vat version of it) here?
```

Comment:
yep, we will, eventually, when we get to cross-machine retirement.

The complicating factor is the async coupling between the two sides: A might tell B that promise 4 has been resolved+retired, but B might have a message in flight that references it, so A can't really forget the promise until it gets some sort of ack from B, which we don't have yet.